### PR TITLE
[Regression] TF-2528 Fix sort by sender name incorrect

### PR DESCRIPTION
**Pull request by dab246**
## Issue

#2528

## Solution

- Sort email list by `Email/query`

## Resolved

- Demo:

https://github.com/user-attachments/assets/370527ff-101a-4be7-92f5-f2a2efcfcdcc

- Integration Test use Provisioning:

[demo-test.webm](https://github.com/user-attachments/assets/73233d31-c2e2-41f6-b9a9-4a9e1b50bf67)

### CODE REVIEW

**Comment by chibenwa** [comment]
Better indeed!

**Review by tddang-linagora — DISMISSED** [review]

File: lib/features/thread/data/network/thread_api.dart [hunk 1]
```diff
@@ -71,16 +75,32 @@ class ThreadAPI {
...
+      QueryEmailResponse.deserialize,
+    );
+
+    List<Email>? emailList;
+
+    if (responseOfGetEmailMethod?.list.isNotEmpty == true &&
+        responseOfQueryEmailMethod?.ids.isNotEmpty == true) {
+      log('ThreadAPI::getAllEmail: QUERY_EMAIL_IDS = ${responseOfQueryEmailMethod?.ids}');
+      final listSortedEmail = responseOfGetEmailMethod!.list
+        .sortingByOrderOfIdList(responseOfQueryEmailMethod!.ids.toList());
```
Comment (hoangdat) [inline lib/features/thread/data/network/thread_api.dart:94]: for the case of `loadMore` we don't need to change anything?. still base on `recievedAt` no need to base on position?

**Review by hoangdat — COMMENTED** [review]
we should have `integration_test` for this change.

File: lib/features/thread/data/network/thread_api.dart (hunk 1 above)
Comment (dab246) [reply lib/features/thread/data/network/thread_api.dart:94]: Yes, We've already handled that. Here we just change the `sort` function to use it via `Email/query`

**Comment by dab246** [comment]
> we should have `integration_test` for this change.

Done

**Review by datdang1995 — APPROVED** [review]

**Review by tddang-linagora — DISMISSED** [review]

File: integration_test/robots/search_robot.dart [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    await $(#search_email_text_field).enterText(queryString);
+    await ensureViewVisible($(#suggestion_search_list_view));
+  }
+
+  Future<void> scrollToEndListSearchFilter() async {
+    await $.scrollUntilVisible(
+      finder: $(#mobile_sortBy_search_filter_button),
+      view: $(#search_filter_list_view),
+      scrollDirection: AxisDirection.right,
+      delta: 300,
```
Comment (hoangdat) [inline integration_test/robots/search_robot.dart:21]: can we have another solution to not depend on fixed number?

**Comment by tddang-linagora** [comment]
- [ ] Test failed

<details>
<summary>Logs</summary>

```java
java.lang.AssertionError: Dart test failed: tests.search.search_email_with_sort_order_test__Should__see__list__email__displayed__by__sort__order__selected__when__search__email__successfully
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞═════════════════
The following TestFailure was thrown running a test:
Expected: 'charlotte@example.com'
Actual: 'david@example.com'
Which: is different.
Expected: charlotte@ ...
Actual: david@exam ...
^
Differ at offset 0

When the exception was thrown, this was the stack:
#4      SearchRobot.expectEmailListSortedCorrectBySenderAscending (file:///Users/dangdat/Desktop/dev/linagora-master/tmail-flutter/integration_test/robots/search_robot.dart:52:7)
#5      SearchEmailWithSortOrderScenario.execute (file:///Users/dangdat/Desktop/dev/linagora-master/tmail-flutter/integration_test/scenarios/search_email_with_sort_order_scenario.dart:39:23)
<asynchronous suspension>
#6      main.<anonymous closure> (file:///Users/dangdat/Desktop/dev/linagora-master/tmail-flutter/integration_test/tests/search/search_email_with_sort_order_test.dart:32:7)
<asynchronous suspension>
#7      TestBase.runPatrolTest.<anonymous closure> (file:///Users/dangdat/Desktop/dev/linagora-master/tmail-flutter/integration_test/base/test_base.dart:28:7)
<asynchronous suspension>
#8      patrolTest.<anonymous closure> (package:patrol/src/common.dart:136:7)
<asynchronous suspension>
#9      testWidgets.<anonymous closure>.<anonymous closure> (package:flutter_test/src/widget_tester.dart:183:15)
<asynchronous suspension>
#10     TestWidgetsFlutterBinding._runTestBody (package:flutter_test/src/binding.dart:1017:5)
<asynchronous suspension>
#11     TestWidgetsFlutterBinding._createTestCompletionHandler.<anonymous closure> (package:flutter_test/src/binding.dart:809:12)
<asynchronous suspension>

This was caught by the test expectation on the following line:
file:///Users/dangdat/Desktop/dev/linagora-master/tmail-flutter/integration_test/robots/search_robot.dart line 52
The test description was:
Should see list email displayed by sort order selected when
search email successfully
═════════════════════════════════════════════════════════════════

at pl.leancode.patrol.PatrolJUnitRunner.runDartTest(PatrolJUnitRunner.java:150)
at com.linagora.android.tmail.MainActivityTest.runDartTest(MainActivityTest.java:31)
```

</details>

File: integration_test/robots/search_robot.dart [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    await ensureViewVisible($(#sort_filter_context_menu));
+  }
+
+  Future<void> selectSortOrder(String sortOrderName) async {
+    await $(find.text(sortOrderName)).tap();
+    await $.pump(const Duration(seconds: 2));
+    await ensureViewVisible($(#search_email_list_notification_listener));
+  }
+
+  Future<void> expectEmailListDisplayedCorrectly({
```
Comment (hoangdat) [inline integration_test/robots/search_robot.dart:37]: IMO, move all `expect` statement to specific scenario.

**Review by hoangdat — DISMISSED** [review]

File: integration_test/robots/search_robot.dart (hunk 3 above)
Comment (dab246) [reply integration_test/robots/search_robot.dart]: Done

**Comment by dab246** [comment]
- [x] Build integration test on CI success

Screenshot 2024-11-13 at 16 15 15

**Review by tddang-linagora — APPROVED** [review]

**Review by hoangdat — APPROVED** [review]

**Pull request merged**